File: engine/src/flutter/lib/gpu/binding_set.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/lib/gpu/binding_set.h"

#include <utility>

#include "flutter/lib/gpu/formats.h"
#include "impeller/core/sampler_descriptor.h"
#include "impeller/renderer/sampler_library.h"

namespace flutter {
namespace gpu {

IMPLEMENT_WRAPPERTYPEINFO(flutter_gpu, BindingSet);

BindingSet::BindingSet() = default;

BindingSet::~BindingSet() = default;

// Whether a shader stage can take resource bindings from a render pass.
static bool IsRenderStage(impeller::ShaderStage stage) {
  return stage == impeller::ShaderStage::kVertex ||
         stage == impeller::ShaderStage::kFragment;
}

void BindingSet::RetainShader(Shader& shader) {
  for (const auto& retained : shaders_) {
    if (retained.get() == &shader) {
      return;
    }
  }
  shaders_.push_back(fml::RefPtr<Shader>(&shader));
}

bool BindingSet::AddUniform(
    Shader& shader,
    int uniform_struct_index,
    const std::shared_ptr<const impeller::DeviceBuffer>& buffer,
    size_t offset_in_bytes,
    size_t length_in_bytes) {
  if (!IsRenderStage(shader.GetShaderStage())) {
    return false;
  }
  const Shader::UniformBinding* uniform_struct =
      shader.GetUniformStructAt(uniform_struct_index);
  if (!uniform_struct) {
    return false;
  }
  if (!buffer || offset_in_bytes + length_in_bytes >
                     buffer->GetDeviceBufferDescriptor().size) {
    return false;
  }

  RetainShader(shader);
  buffer_bindings_.push_back(BufferBinding{
      .stage = shader.GetShaderStage(),
      .slot = uniform_struct->slot,
      .metadata = &uniform_struct->metadata,
      .view = impeller::BufferView(
          buffer, impeller::Range(offset_in_bytes, length_in_bytes)),
  });
  return true;
}

bool BindingSet::AddTexture(
    Shader& shader,
    int uniform_texture_index,
    std::shared_ptr<const impeller::Texture> texture,
    impeller::raw_ptr<const impeller::Sampler> sampler) {
  if (!IsRenderStage(shader.GetShaderStage())) {
    return false;
  }
  const Shader::TextureBinding* texture_binding =
      shader.GetUniformTextureAt(uniform_texture_index);
  if (!texture_binding) {
    return false;
  }
  if (!texture || !sampler) {
    return false;
  }

  RetainShader(shader);
  texture_bindings_.push_back(TextureBinding{
      .stage = shader.GetShaderStage(),
      .slot = texture_binding->slot,
      .metadata = &texture_binding->metadata,
      .texture = std::move(texture),
      // NOLINTNEXTLINE(performance-move-const-arg)
      .sampler = std::move(sampler),
  });
  return true;
}

void BindingSet::Clear() {
  buffer_bindings_.clear();
  texture_bindings_.clear();
  shaders_.clear();
}

const std::vector<BindingSet::BufferBinding>& BindingSet::GetBufferBindings()
    const {
  return buffer_bindings_;
}

const std::vector<BindingSet::TextureBinding>& BindingSet::GetTextureBindings()
    const {
  return texture_bindings_;
}

}  // namespace gpu
}  // namespace flutter

//----------------------------------------------------------------------------
/// Exports
///

void InternalFlutterGpu_BindingSet_Initialize(Dart_Handle wrapper) {
  auto res = fml::MakeRefCounted<flutter::gpu::BindingSet>();
  res->AssociateWithDartWrapper(wrapper);
}

bool InternalFlutterGpu_BindingSet_AddUniform(
    flutter::gpu::BindingSet* wrapper,
    flutter::gpu::Shader* shader,
    int uniform_struct_index,
    flutter::gpu::DeviceBuffer* device_buffer,
    int offset_in_bytes,
    int length_in_bytes) {
  if (offset_in_bytes < 0 || length_in_bytes < 0) {
    return false;
  }
  return wrapper->AddUniform(*shader, uniform_struct_index,
                             device_buffer->GetBuffer(), offset_in_bytes,
                             length_in_bytes);
}

bool InternalFlutterGpu_BindingSet_AddTexture(flutter::gpu::BindingSet* wrapper,
                                              flutter::gpu::Context* context,
                                              flutter::gpu::Shader* shader,
                                              int uniform_texture_index,
                                              flutter::gpu::Texture* texture,
                                              int min_filter,
                                              int mag_filter,
                                              int mip_filter,
                                              int width_address_mode,
                                              int height_address_mode,
                                              int max_anisotropy) {
  const impeller::SamplerDescriptor sampler_desc =
      flutter::gpu::ToImpellerSamplerDescriptor(
          min_filter, mag_filter, mip_filter, width_address_mode,
          height_address_mode, max_anisotropy);
  return wrapper->AddTexture(
      *shader, uniform_texture_index, texture->GetTexture(),
      context->GetContext().GetSamplerLibrary()->GetSampler(sampler_desc));
}

void InternalFlutterGpu_BindingSet_Clear(flutter::gpu::BindingSet* wrapper) {
  wrapper->Clear();
}

File: engine/src/flutter/lib/gpu/lib/src/binding_set.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// ignore_for_file: public_member_api_docs

part of flutter_gpu;

/// A texture and the sampler configuration used to read it, for use in a
/// [BindingSet].
base class TextureBinding {
  const TextureBinding(this.texture, {this.sampler});

  /// The texture to read.
  final Texture texture;

  /// The sampler to read [texture] with. Defaults to the same nearest
  /// filtering and clamped addressing as [RenderPass.bindTexture] when null.
  final SamplerOptions? sampler;
}

/// An immutable group of uniform and texture bindings, resolved against
/// shader reflection once at creation and reusable across draws and passes.
///
/// Create one with [GpuContext.createBindingSet] and bind it with
/// [RenderPass.bindSet]. Binding costs one slot assignment no matter how many
/// resources the set holds, so a renderer that draws one material across many
/// nodes stops paying per-resource bind work on every draw.
///
/// A set references its resources rather than copying them, so it stays valid
/// for as long as the bound buffers and textures do. Data that changes from
/// frame to frame produces a new [BufferView] each frame (see
/// [HostBuffer.emplace]) and belongs in a per-draw [RenderPass.bindUniform]
/// instead.
base class BindingSet extends NativeFieldWrapperClass1 {
  BindingSet._(
    this._gpuContext,
    Map<UniformSlot, BufferView> uniforms,
    Map<UniformSlot, TextureBinding> textures,
  ) : _uniforms = Map<UniformSlot, BufferView>.unmodifiable(uniforms),
      _textures = Map<UniformSlot, TextureBinding>.unmodifiable(textures) {
    _initialize();
    _populate();
  }

  final GpuContext _gpuContext;
  final Map<UniformSlot, BufferView> _uniforms;
  final Map<UniformSlot, TextureBinding> _textures;

  /// The shader reload epoch the native bindings were resolved against. A
  /// reload replaces the reflection data they point at, so the set is
  /// repopulated the next time it is bound.
  int _epoch = -1;

  /// Resolves every entry against shader reflection and hands it to the
  /// native set. Throws for a name the shader does not declare, a buffer view
  /// that runs past the end of its buffer, or an invalid sampler.
  void _populate() {
    for (final MapEntry<UniformSlot, BufferView> entry in _uniforms.entries) {
      final UniformSlot slot = entry.key;
      final int structIndex = slot._resolvedStructIndex;
      if (structIndex < 0) {
        throw Exception(
          "Failed to bind uniform (no uniform struct named '${slot.uniformName}')",
        );
      }
      final BufferView view = entry.value;
      if (!_addUniform(
        slot.shader,
        structIndex,
        view.buffer,
        view.offsetInBytes,
        view.lengthInBytes,
      )) {
        throw Exception("Failed to bind uniform '${slot.uniformName}'");
      }
    }

    for (final MapEntry<UniformSlot, TextureBinding> entry
        in _textures.entries) {
      final UniformSlot slot = entry.key;
      final int textureIndex = slot._resolvedTextureIndex;
      if (textureIndex < 0) {
        throw Exception(
          "Failed to bind texture (no texture named '${slot.uniformName}')",
        );
      }
      final TextureBinding binding = entry.value;
      final SamplerOptions sampler = binding.sampler ?? SamplerOptions();
      _validateBindableTexture(binding.texture);
      sampler._validate();
      if (!_addTexture(
        _gpuContext,
        slot.shader,
        textureIndex,
        binding.texture,
        sampler.minFilter.index,
        sampler.magFilter.index,
        sampler.mipFilter.index,
        sampler.widthAddressMode.index,
        sampler.heightAddressMode.index,
        sampler.maxAnisotropy,
      )) {
        throw Exception("Failed to bind texture '${slot.uniformName}'");
      }
    }

    _epoch = _shaderReloadEpoch;
  }

  /// Re-resolves the set after a shader hot reload. Cheap no-op otherwise.
  void _syncReloadEpoch() {
    if (_epoch == _shaderReloadEpoch) {
      return;
    }
    _clear();
    _populate();
  }

  /// Wrap with native counterpart.
  @Native<Void Function(Handle)>(
    symbol: 'InternalFlutterGpu_BindingSet_Initialize',
  )
  external void _initialize();

  @Native<
    Bool Function(Pointer<Void>, Pointer<Void>, Int, Pointer<Void>, Int, Int)
  >(symbol: 'InternalFlutterGpu_BindingSet_AddUniform')
  external bool _addUniform(
    Shader shader,
    int uniformStructIndex,
    DeviceBuffer buffer,
    int offsetInBytes,
    int lengthInBytes,
  );

  @Native<
    Bool Function(
      Pointer<Void>,
      Pointer<Void>,
      Pointer<Void>,
      Int,
      Pointer<Void>,
      Int,
      Int,
      Int,
      Int,
      Int,
      Int,
    )
  >(symbol: 'InternalFlutterGpu_BindingSet_AddTexture')
  external bool _addTexture(
    GpuContext gpuContext,
    Shader shader,
    int uniformTextureIndex,
    Texture texture,
    int minFilter,
    int magFilter,
    int mipFilter,
    int widthAddressMode,
    int heightAddressMode,
    int maxAnisotropy,
  );

  @Native<Void Function(Pointer<Void>)>(
    symbol: 'InternalFlutterGpu_BindingSet_Clear',
  )
  external void _clear();
}
